File: Algorithm/Python/746. Min Cost Climbing Stairs.py
# -*- coding: utf-8 -*-
# Created on Sun Dec 17 2017 11:28:47
# Author: WuLC
# EMail: liangchaowu5@gmail.com


# dp, O(1) space
class Solution(object):
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """
        last1, last2 = cost[0], cost[1]
        for i in range(2, len(cost)):
            curr = min(last1, last2) + cost[i]
            last1, last2 = last2, curr
        return min(last1, last2)
            
        

File: Algorithm/C++/746. Min Cost Climbing Stairs.cpp
/*
 * Created on Wed May 02 2018 10:39:32
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// dp, O(n) space
// pay attention that top means exceed the last element of the array
class Solution 
{
    public:
        int minCostClimbingStairs(vector<int>& cost)
        {
            std::vector<int> dp = {0,0};
            for(int i=2; i<=cost.size(); i++)
                dp.push_back(std::min(dp[i-1]+cost[i-1], dp[i-2]+cost[i-2]));
            return dp.back();
        }
};